Adding functionalities for QuadraturePointMultiapp system

abarun22 opened the pull request:
This PR implements a new feature that runs a multiapp system based on elemental quadrature points

## Reason
High fidelity multi-scale models requires probing at micro-scales and very often require a computational treatment on their representative volume elements and may require homogenized quantities to be transferred to the quadrature points of the model at the macro scale. This funtionality would serve as a multiapp framework that will create a sub-app at each of the spatial locations of the macro quadrature points and forms the basis for such computational models.

Example tests are created to demonstrate the purpose of this multiapp system.

GiudGiud commented:
@abarun22 the diffs in MultiApp.h and MultiApp.C are very very large, seemingly excessive. How much did you modify these files? Or is this a git issue that it's finding too much diff?

abarun22 commented:
> @abarun22 the diffs in MultiApp.h and MultiApp.C are very very large, seemingly excessive. How much did you modify these files? Or is this a git issue that it's finding too much diff?

That's quite strange. I am pretty sure the modifications are minimal. These are mainly related to vectorizing the `_reset_time` variable which you observe in the `validparams`, `constructor` and the function `preTransfer`. It should definitely be due to other issues.

abarun22 commented:
> @abarun22 the diffs in MultiApp.h and MultiApp.C are very very large, seemingly excessive. How much did you modify these files? Or is this a git issue that it's finding too much diff?

Eager to get this PR done. Please let me know if there are any difficulties in merging this code with the repo.

GiudGiud commented:
Well first let's separate the work on reset times and on quadrature multiapp in two PRs. This diff issue needs to be figured out as well. You are getting merge conflicts right now because of it. 

You also need to create an issue in github (more of a SQA formality at this point) to discuss the feature request, for both features separately. 

The testing is insufficient. Where the transfers and multiapps systems are currently at in terms of reliability, we would want to test every single transfer (roughly 10 in the framework/, not every single everywhere) with this new MultiApp.

abarun22 commented:
> Well first let's separate the work on reset times and on quadrature multiapp in two PRs. This diff issue needs to be figured out as well. You are getting merge conflicts right now because of it.
> 
> You also need to create an issue in github (more of a SQA formality at this point) to discuss the feature request, for both features separately.
> 
> The testing is insufficient. Where the transfers and multiapps systems are currently at in terms of reliability, we would want to test every single transfer (roughly 10 in the framework/, not every single everywhere) with this new MultiApp. We need to see that all the ones that are relevant work properly, and the ones that are not error gracefully basically

As advised separate issues has been created each for "reset times" and "Quadrature point multiapp". I see that there are 24 transfer schemes in place in the framework and i wonder whether you want me to test with all of them. This would really be a time consuming task if we have to do for all of the transfer methods.

abarun22 commented:
> > Well first let's separate the work on reset times and on quadrature multiapp in two PRs. This diff issue needs to be figured out as well. You are getting merge conflicts right now because of it.
> > You also need to create an issue in github (more of a SQA formality at this point) to discuss the feature request, for both features separately.
> > The testing is insufficient. Where the transfers and multiapps systems are currently at in terms of reliability, we would want to test every single transfer (roughly 10 in the framework/, not every single everywhere) with this new MultiApp. We need to see that all the ones that are relevant work properly, and the ones that are not error gracefully basically
> 
> As advised separate issues has been created each for "reset times" and "Quadrature point multiapp". I see that there are 24 transfer schemes in place in the framework and i wonder whether you want me to test with all of them. This would really be a time consuming task if we have to do for all of the transfer methods.

These are the issue ID's that were created: #22490 and #22489

abarun22 commented:
The new commits are now under the PR refs: #22529. Mistakenly i have included the files for QuadraturePointMultiapp in this new PR. You could trace these files under the commit message refs: #22490. Feel free to ignore these two files (.C and .h) while merging in to your main repo. 
Regarding testing, i believe you want me to prepare 24 tests for all the transfers occuring in combination with this new multiapp?

GiudGiud commented:
@abarun22 please just make one test and I ll do the extension to all the other transfers.

You'll need to fix the other PR to only include reset-related items. 
Also you have a merge conflict

abarun22 commented:
> @abarun22 please just make one test and I ll do the extension to all the other transfers.
> 
> You'll need to fix the other PR to only include reset-related items. Also you have a merge conflict so you need to rebase on devel

I guess you are asking me to remove QuadraturePointMultiapp (.C and .h) and push it on to the remote enroute the same PR. I think this could be the simplest method to amend the problematic PR. I do not understand what you meant by saying a rebase on "devel"

abarun22 commented:
I performed the rebasing operation as you suggested and the sources are now in sync with the recent moose updates. The sources for reset times alone are now committed under a new pull request https://github.com/idaholab/moose/pull/22582. I assume that you are OK with the tests under the commit "Add tests for QuadraturePointMultiapp" (https://github.com/idaholab/moose/commit/48cae5e6622b07ade47d834a09ee16b658d98813). Please let me know if yo need more, i am happy to share with you.

abarun22 commented:
A correction please. Due to some issues i had to close that PR #22582. The new one can now be traced here https://github.com/idaholab/moose/pull/22583

GiudGiud commented:
Ok good for the separate PR. I ll go help there when I have time.

For this PR, the current test does not work.
- the multiapp does not do much in the simulation since there is no transfer of information back to the main app
- it's written as a mechanics problem, but this is a framework capability and you need to only use framework object. I d recommend writing a simple diffusion problem instead

GiudGiud reviewed (dismissed):
see comment above

GiudGiud committed:
Add a multiapp option that automaticaly generates positions at quadrature points, refs #22490

Co-authored-by: Arun P Balasubramanian <arun.balasubramanian@ukaea.uk>

GiudGiud committed:
Remove duplicate quadrature points at the boundaries between elements
refs #22490

GiudGiud committed:
Add test for quadrature point multiapp involving transfers back and forth of constant data
refs #22490

GiudGiud committed:
Fix multiapp interpolation of posptorcessor transfer for the case where the app is running in position, refs #19121

GiudGiud committed:
Fix multiapp user object transfer obtention of app position for the case where the app is running in position, refs #19121

GiudGiud committed:
Add documentation for new Qp multiapp refs #22490

GiudGiud commented:
@abarun22 I apologize profusely for the delay in helping out here. This is a pretty awesome functionality and we'll get some use for it for high fidelity simulations. 

I added a test and removed the large diffs. 
You are a co-author on the main commit adding this functionality.

@lindsayad for the review. 
I tagged on two fixes for transfers for apps running in position. Running the subapps in position helps make easy tests that dont require any diffusion solve.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1382027/) on 219cbea wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21340/site)

This comment will be updated on new commits.

lindsayad commented on framework/doc/content/source/multiapps/QuadraturePointMultiApp.md:
```diff
@@ -0,0 +1,26 @@
+be generated on specified subdomains.
+
+!alert note
+Quadrature points that are shared between elements are collapsed into the same sub-application. Please consider the
```
Looking at the source, it looks like you are only looking at interior quadrature points anyway

GiudGiud replied:
I ll adapt the docstring

lindsayad replied:
There are two types of default quadrature rules, `QClough` and `QGauss`. Gaussian quadrature points should never be on the element boundaries (https://en.wikipedia.org/wiki/Gaussian_quadrature). Looking through the source it looks like `QClough` uses linear combinations of Gauss points which I'm guessing are still all on the interior (@roystgnr maybe could confirm).

Even a non-default trapezoidal quadrature would have quadrature points at element boundaries but you are iterating over elements and then asking for the quadrature points from that element ... so you would have duplicate points in your `_positions` vector in this case. Maybe this is what you mean by "collapsed"?

Anyway given that default quadrature will probably always be Gauss for MOOSE, I think we can delete this comment now especially that you've added the comment below.

GiudGiud replied:
sounds good. 
By collapsed i mean we remove the duplicates.
Anyhow it s removed now

roystgnr replied:
`QClough` points are all interior; it's basically just a Gaussian rule on each subtriangle inside the macroelement.

We do have more than just `QTrap` giving points on element boundaries: `QGaussLobatto`, `QSimpson`, `QNodal`.  We never get any of those as a default quadrature rule out of a libMesh `FEMContext` or anything, though; Clough-Tocher variables are the only things that don't play well enough with Gauss that we have to account for them.

FIN-S used IIRC `QNodal`, so that all the godawful expensive transcendental-function-filled thermochemical models could be evaluated fewer times (sharing results between elements), but I haven't heard of MOOSE apps doing anything like that.

lindsayad commented on framework/src/multiapps/QuadraturePointMultiApp.C:
```diff
@@ -0,0 +1,87 @@
+    // FEType is local to the element, supporting mixed order meshes
+    const FEFamily mapping_family = FEMap::map_fe_type(*elem);
+    FEType fe_type(elem->default_order(), mapping_family);
+    const auto qrule = fe_type.default_quadrature_rule(elem->dim());
```
If you do, you'd want to make some of the qrule accessors in `Assembly` `public` and then query for the order of the volumetric rule

GiudGiud replied:
I thought I replied here

GiudGiud replied:
anyhow, I ll just document it. 
I dont want to go modify Assembly for this, 1st and second order being covered well should do all that we need here

lindsayad replied:
works for me

lindsayad commented on test/tests/multiapps/quadrature_point_multiapp/sub_app.i:
```diff
@@ -0,0 +1,56 @@
+
+[Executioner]
+  type = Transient
+  num_steps = 1
+  petsc_options_iname = '-pc_type -pc_hypre_type'
+  petsc_options_value = 'hypre boomeramg'
+  solve_type = PJFNK
```
```suggestion
```
if we're not solving anything

lindsayad commented on test/tests/multiapps/quadrature_point_multiapp/quadrature_point_multiapp.i:
```diff
@@ -0,0 +1,91 @@
+
+[Executioner]
+  type = Transient
+  num_steps = 1
+  petsc_options_iname = '-pc_type -pc_hypre_type'
+  petsc_options_value = 'hypre boomeramg'
+  solve_type = NEWTON
```
```suggestion
```
if we're not solving anything

lindsayad commented on test/tests/multiapps/quadrature_point_multiapp/tests:
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues = '#22490'
+  design = 'QuadraturePointMultiApp.md'
+
```
```suggestion
```

GiudGiud committed:
Apply suggestions from code review
- remove extra includes
- dont set solvers if you re not solving in tests
- be more explicit about limitations in documentation

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Test failure is unrelated.

Thanks for the contribution @abarun22 !
